akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala

```diff
@@ -4,46 +4,151 @@
 package akka.stream.impl.io
 
 import java.io.InputStream
-import java.nio.file.Path
+import java.nio.ByteBuffer
+import java.nio.channels.{ AsynchronousFileChannel, CompletionHandler }
+import java.nio.file.{ Files, Path, StandardOpenOption }
 
+import akka.Done
+import akka.stream.Attributes.InputBuffer
 import akka.stream._
 import akka.stream.ActorAttributes.Dispatcher
 import akka.stream.IOResult
+import akka.stream.impl.Stages.DefaultAttributes
 import akka.stream.impl.StreamLayout.Module
 import akka.stream.impl.Stages.DefaultAttributes.IODispatcher
 import akka.stream.impl.{ ErrorPublisher, SourceModule }
+import akka.stream.stage._
 import akka.util.ByteString
 import org.reactivestreams._
+
 import scala.concurrent.{ Future, Promise }
+import scala.util.{ Failure, Success, Try }
 
 /**
  * INTERNAL API
- * Creates simple synchronous Source backed by the given file.
  */
-private[akka] final class FileSource(f: Path, chunkSize: Int, val attributes: Attributes, shape: SourceShape[ByteString])
-  extends SourceModule[ByteString, Future[IOResult]](shape) {
-  require(chunkSize > 0, "chunkSize must be greater than 0")
-  override def create(context: MaterializationContext) = {
-    // FIXME rewrite to be based on GraphStage rather than dangerous downcasts
-    val materializer = ActorMaterializerHelper.downcast(context.materializer)
-    val settings = materializer.effectiveSettings(context.effectiveAttributes)
+private[akka] object FileSource {
 
-    val ioResultPromise = Promise[IOResult]()
-    val props = FilePublisher.props(f, ioResultPromise, chunkSize, settings.initialInputBufferSize, settings.maxInputBufferSize)
-    val dispatcher = context.effectiveAttributes.get[Dispatcher](IODispatcher).dispatcher
+  val completionHandler = new CompletionHandler[Integer, Try[Int] ⇒ Unit] {
 
-    val ref = materializer.actorOf(context, props.withDispatcher(dispatcher))
+    override def completed(result: Integer, attachment: Try[Int] ⇒ Unit): Unit = {
+      attachment(Success(result))
+    }
 
-    (akka.stream.actor.ActorPublisher[ByteString](ref), ioResultPromise.future)
+    override def failed(ex: Throwable, attachment: Try[Int] ⇒ Unit): Unit = {
+      attachment(Failure(ex))
+    }
   }
+}
 
-  override protected def newInstance(shape: SourceShape[ByteString]): SourceModule[ByteString, Future[IOResult]] =
-    new FileSource(f, chunkSize, attributes, shape)
+/**
+ * INTERNAL API
+ * Creates simple asynchronous Source backed by the given file.
+ */
+private[akka] final class FileSource(path: Path, chunkSize: Int)
+  extends GraphStageWithMaterializedValue[SourceShape[ByteString], Future[IOResult]] {
+  require(chunkSize > 0, "chunkSize must be greater than 0")
 
-  override def withAttributes(attr: Attributes): Module =
-    new FileSource(f, chunkSize, attr, amendShape(attr))
+  val out = Outlet[ByteString]("FileSource.out")
+
+  override protected def initialAttributes: Attributes = DefaultAttributes.fileSource
```

Comment:
This matches my previous benchmarks and observations when I was imlpementing the initial one. On raw throughput the blocking IO with dedicated dispatcher is faster. I wonder if this will scale better for many file access though.